### PR TITLE
Add CC-BY-4.0 attribution for Red Hat-sourced advisory content

### PR DESCRIPTION
**Pull request by rockythorn**
## Problem

Rocky Linux errata advisories (e.g. RLSA-2026:23102) re-publish content from the
corresponding Red Hat advisories — synopsis, description, and topic. Red Hat
licenses that advisory content under [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/),
which permits reuse but requires crediting the source, linking to the original
advisory and to the license, and indicating that changes were made. None of
Apollo's published outputs currently do this. This was raised by a member of the
Rocky community.

## Approach

Attribution text and links are centralized in a new `apollo/server/attribution.py`
and applied to every surface that carries Red Hat-derived text:

- **updateinfo.xml** — the `<rights>` element credits the source advisory under
  CC BY 4.0, plus `<reference>` entries for the source advisory (`type="vendor"`)
  and the license (`type="other"`).
- **v2 and v3 JSON APIs** — a structured `source` object: `name`, `url`,
  `vendor`, `license`, `licenseUrl`.
- **OSV** — a source `ADVISORY` reference, a Red Hat credit, and
  `license`/`license_url`/`source_advisory` in `database_specific`.
- **RSS feed** — the channel copyright carries the CC BY 4.0 notice.
- **Server-rendered advisory page** — an attribution line linking the source
  advisory and the license.

No database migration is required; the source advisory and its URL are derived
from the existing `red_hat_advisory` relation. A plain copyright line is kept as
a fallback for any advisory without a Red Hat source.

## Testing

- New unit tests across updateinfo, OSV, and v2/v3 serialization; all Bazel
  suites pass.
- Validated against a development database with the server running: v2 and v3
  detail + list, OSV, the live `updateinfo.xml` (889 updates in one repo file),
  RSS, and the web advisory page all carry the attribution.
- Verified `dnf` tolerates the new `updateinfo` reference types, both at the
  source level (libsolv/libdnf/libdnf5) and end-to-end via `dnf makecache` on
  Rocky 9.

## Notes

- Companion change in `rocky-linux/errata-frontend` (branch
  `feature/rhel-advisory-attribution`) renders the per-advisory attribution and
  a site-wide footer notice.
- The exact attribution wording is under review; prose may be adjusted.
- No new dependencies.

### CODE REVIEW

**Comment by rockythorn** [comment]
Companion frontend change that renders this attribution: rocky-linux/errata-frontend#4

File: apollo/server/routes/api_advisories.py [hunk 1]
```diff
@@ -37,15 +52,18 @@ async def list_advisories(
...
-    advisories = await paginate(
-        Advisory.all().prefetch_related(
-            "red_hat_advisory",
-            "packages",
-            "cves",
-            "fixes",
-            "affected_products",
-        ).order_by("-published_at"),
-    )
+    query = Advisory.all().prefetch_related(
```
Comment (jdieter) [inline apollo/server/routes/api_advisories.py:60]: Do we have any tests covering this?

File: apollo/server/routes/api_updateinfo.py [hunk 2]
```diff
@@ -151,9 +152,13 @@ def generate_updateinfo_xml(
...
         now = datetime.datetime.utcnow()
-        ET.SubElement(
-            update, "rights"
-        ).text = f"Copyright {now.year} {company_name}"
+        if advisory.red_hat_advisory_id:
+            rights_text = attribution.attribution_rights(
+                advisory.red_hat_advisory.name, company_name, now.year
+            )
+        else:
+            rights_text = f"Copyright {now.year} {company_name}"
```
Comment (jdieter) [inline apollo/server/routes/api_updateinfo.py:160]: Can red_hat_advisory_id be null?  If not, then the else statement can never be called.  This also means that we can only ever issue advisories that originate from RH, though fixing that's a task for another PR.

File: apollo/server/templates/advisory.jinja [hunk 3]
```diff
@@ -35,6 +35,14 @@
     </h5>
     <h5 style="font-weight:500">Updated at: <span style="font-weight:400;">{{ advisory.updated_at.date() }}</span></h5>
   </div>
+  {% if advisory.red_hat_advisory_id %}
```
Comment (jdieter) [inline apollo/server/templates/advisory.jinja:38]: There's a lot of hardcoded stuff here that we also have added to attribution.py.  We can't import attribution.py here, so maybe add a comment to both so developers know they have to keep them in sync?

File: apollo/server/routes/api_advisories.py (hunk 1 above)
Comment (rockythorn) [reply apollo/server/routes/api_advisories.py:60]: Added `apollo/tests/test_api_advisories.py` covering `_advisory_source` (the helper both the list and detail endpoints use): it asserts the structured `source` is built from the Red Hat advisory when present and is `None` when `red_hat_advisory_id` is absent. `attribution.source_fields` and the serializer models are also covered in `test_api_compat.py`.

File: apollo/server/routes/api_updateinfo.py (hunk 2 above)
Comment (rockythorn) [reply apollo/server/routes/api_updateinfo.py:160]: Good question — I checked the actual schema rather than the ORM model. `red_hat_advisory_id` is nullable: `schema.sql` declares it `bigint` with no NOT NULL, and the live DB reports `is_nullable = YES` (the FK allows NULL and is ON DELETE CASCADE). The Tortoise model marks the FK non-nullable, but that is not enforced in the database, so the `else` is not dead code — it keeps a single null row from invalidating the entire `updateinfo.xml` for a repo, so I have left it in place. The ORM/DB nullability mismatch is probably worth tightening separately, and supporting genuinely non-RH advisories would be its own change, as you note.

File: apollo/server/templates/advisory.jinja (hunk 3 above)
Comment (rockythorn) [reply apollo/server/templates/advisory.jinja:38]: Made `attribution.py` the single source of truth: it is now registered as a Jinja global (`templates.env.globals["attribution"]` in `apollo/server/utils.py`), and the template references `{{ attribution.red_hat_errata_url(...) }}`, `{{ attribution.SOURCE_VENDOR }}`, `{{ attribution.SOURCE_LICENSE_URL }}`, and `{{ attribution.LICENSE_NAME }}` instead of hardcoded values — so there is nothing to keep in sync. Verified the rendered page is unchanged.

**Comment by rockythorn** [comment]
Revised in response to the review comments here, plus a separate pass on the attribution wording. Everything is folded into the existing single commit (force-pushed).

**Addressing the review**

- **Tests** — added `apollo/tests/test_api_advisories.py` covering the v3 `source` derivation (both the Red Hat-sourced path and the no-source path). `attribution.source_fields` and the serializer models are also covered in `apollo/tests/test_api_compat.py`.
- **`red_hat_advisory_id` nullability** — I checked the actual schema rather than the ORM: the column is nullable (`schema.sql` declares it `bigint` with no `NOT NULL`, and the live DB reports `is_nullable = YES`; the FK permits NULL). The Tortoise model marks the FK non-nullable, but that isn't enforced in the database, so the fallback isn't dead — it keeps a single null row from invalidating an entire `updateinfo.xml`. I left that thread open in case you'd rather tighten the schema (add `NOT NULL`) instead, which would let us drop the fallback.
- **Hardcoded values in `advisory.jinja`** — `attribution.py` is now registered as a Jinja global (`apollo/server/utils.py`), and the template references `{{ attribution.red_hat_errata_url(...) }}` / `{{ attribution.SOURCE_VENDOR }}` / `{{ attribution.SOURCE_LICENSE_URL }}` / `{{ attribution.LICENSE_NAME }}` instead of hardcoded strings, so there is a single source of truth.

**Wording pass** (separate review of the attribution text)

- The copyright notice uses the `©` symbol instead of `(C)`.
- Standardized the change indication on "…, with modifications." (previously "…; modified by <company>").
- Both are centralized in `apollo/server/attribution.py`.

All 14 Bazel test suites pass, and the change was validated end-to-end against a development database — the v2 and v3 advisory APIs, OSV, the live `updateinfo.xml`, RSS, and the rendered advisory page all carry the attribution.

File: apollo/server/routes/api_advisories.py (hunk 1 above)
Comment (jdieter) [reply apollo/server/routes/api_advisories.py:60]: Sorry, I should have been clearer.  We're switching from using fastapi pagination to our own home-grown pagination.  Can we please have a test that verifies that this pagination works as expected?

Comment (rockythorn) [reply apollo/server/routes/api_advisories.py:60]: Understood — added `TestListAdvisoriesPagination` to `apollo/tests/test_api_advisories.py`, which covers the hand-rolled pagination specifically.

It drives `list_advisories` through a `TestClient` on a minimal app with `add_pagination(app)`, so the real `Params` query parsing and the `create_page` path both run. A fake queryset stands in for Tortoise: `offset`/`limit` return a new queryset, awaiting it yields the slice, and `count()` reports the unsliced size — over a fixed list of 25 rows.

What it asserts:

- Default params → `page=1`, `size=50`, all 25 items.
- `page=1&size=10` → rows 1–10; `page=2&size=10` → rows 11–20 (i.e. offset really is `size * (page - 1)`).
- `page=3&size=10` → the 5-item partial final page.
- `page=4&size=10` → empty `advisories`, with `total` still 25.
- `total` is the full result-set count, not the length of the page.
- Paging through with `size=7` returns all 25 advisories, in order, exactly once — no gaps or overlaps.
- `last_updated_at` still comes from `RedHatIndexState`.

I also checked the tests fail on regressions rather than passing vacuously: an off-by-one offset (`size * page`) fails 5 of them, dropping `.limit(params.size)` fails 4, and computing `total = len(page_orm)` fails 2.

One build change: `apollo/tests/BUILD.bazel` now lists `@pypi_httpx//:pkg` for `fastapi.testclient`. That isn't a new dependency — httpx 0.23.3 is already in `requirements_lock.txt` (via `openapi-python-client`); the target just needs it declared. `bazel test apollo/tests:all` passes (14/14).

**Review by jdieter — APPROVED** [review]
LGTM
